azure-core/azure/core/exceptions.py

```diff
@@ -0,0 +1,148 @@
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+# --------------------------------------------------------------------------
+
+import logging
+import sys
+
+from typing import Callable, Any, Optional, TYPE_CHECKING
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def raise_with_traceback(exception, message="", *args, **kwargs):
+    # type: (Callable, str, Any, Any) -> None
+    """Raise exception with a specified traceback.
+
+    This MUST be called inside a "except" clause.
+
+    :param Exception exception: Error type to be raised.
+    :param str message: Message to include with error, empty by default.
+    :param args: Any additional args to be included with exception.
+    """
+    exc_type, exc_value, exc_traceback = sys.exc_info()
+    # If not called inside a "except", exc_type will be None. Assume it will not happen
+    exc_msg = "{}, {}: {}".format(message, exc_type.__name__, exc_value)  # type: ignore
+    error = exception(exc_msg, *args, **kwargs)
+    try:
+        raise error.with_traceback(exc_traceback)
+    except AttributeError:
+        error.__traceback__ = exc_traceback
+        raise error
+
+
+class AzureError(Exception):
+    """Base exception for all errors."""
+
+    def __init__(self, message, *args, **kwargs):
+        self.inner_exception = kwargs.get('error')
+        self.response = kwargs.get('response')
+        self.exc_type, self.exc_value, self.exc_traceback = sys.exc_info()
+        self.exc_type = self.exc_type.__name__ if self.exc_type else type(self.inner_exception)
+        self.exc_msg = "{}, {}: {}".format(message, self.exc_type, self.exc_value)  # type: ignore
+        self.message = str(message)
+        super(AzureError, self).__init__(self.message, *args)
+
+    def raise_with_traceback(self):
+        try:
+            raise super(AzureError, self).with_traceback(self.exc_traceback)
+        except AttributeError:
+            self.__traceback__ = self.exc_traceback
+            raise self
+
+
+class AzureLibraryError(AzureError):
+    """An error occurred in the client while processing the pipeline."""
+
+
+class AzureLibraryResponse(AzureLibraryError):
+    """An error occurred in the client while processing the response."""
+
+
+class AzureLibraryRequest(AzureLibraryError):
+    """An error occurred in the client while building the request."""
+
+
+class DecodeError(AzureLibraryResponse):
+    """Error raised during response deserialization."""
+
+
+class ServiceRequestError(AzureError):
+    """An error occurred while attempt to make a request to the service."""
+
+
+class ConnectionError(ServiceRequestError):
```

Comment:
That's a builtin exception as well, that doesn't need import. This means if someone writes:
```python
from azure.core.exceptions import ConnectionError
```
He will replace the builtin exception. 
That seems an unnecessary pitfall for me

Reply:
Yeah I'm aware of this. Requests gets around it by using "ConnectError" - I raised this in the Exceptions conversation, so hoping a name will be settled on there:
https://github.com/Azure/azure-sdk/pull/195